api/src/main/java/io/jsonwebtoken/lang/UnavailableImplementationException.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright (C) 2019 jsonwebtoken.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.jsonwebtoken.lang;
+
+/**
+ * Exception indicating that no implementation of an jjwt-api SPI was found on the classpath.
+ */
+public class UnavailableImplementationException extends RuntimeException {
+
+    private static final String DEFAULT_NOT_FOUND_MESSAGE = "Unable to find an implementation for %s using java.util.ServiceLoader. Ensure you include a backing implementation .jar in the classpath, for example jjwt-impl.jar, or your own .jar for custom implementations.";
+
+    public UnavailableImplementationException(final Class klass) {
+        super(String.format(DEFAULT_NOT_FOUND_MESSAGE, klass));
```

Comment:
This might be pedantic, but should we have a `notNull` assertion on `klass`?  Feel free to add it or not - just an observation.

Reply:
I think `null` is a valid case here for a couple of reasons:
1. I'd rather not throw an exception from another exception
2. the assertion message will contain `Unable to find an implementation for %s` which while not extremely useful, is valid.  I think methods that throw this exception should be checking for null implementations.

That said, making the class final, and making this constructor package-private should reduce the possibility of this coming up.

Thoughts?

Reply:
Ah, right - throwing an exception from an exception.  I guess I was in 'no nulls everywhere possible' mode.

I can see both sides - leaving it alone and still including the null assertion.  I'm on the fence but I'd probably pick putting a null assertion in there because it means that the class is being used illegally - the upstream code didn't check for null and that would surface a programming error/bug on our part, which is a good thing I suppose.

That said, have you verified that all upstream callers have checked for null first?  If so, I'm fine with whatever you decide.

Reply:
Just want to bump in case you missed this thread - If not putting in a null assertion, have you been able to verify upstream callers do so?